src/compile/common.ts

```diff
@@ -111,24 +110,29 @@ export function formatSignalRef(fieldDef: FieldDef<string>, specifiedFormat: str
  *
  * @param format explicitly specified format
  */
-export function numberFormat(fieldDef: FieldDef<string>, specifiedFormat: string, config: Config, channel: Channel) {
-  // Specified format in axis/legend has higher precedence than fieldDef.format
-  const format = specifiedFormat;
+export function numberFormat(fieldDef: FieldDef<string>, specifiedFormat: string, config: Config) {
   if (fieldDef.type === QUANTITATIVE) {
     // add number format for quantitative type only
 
-    if (format) {
-      return format;
-    } else if (fieldDef.aggregate === 'count' && channel === TEXT) {
-      // FIXME: need a more holistic way to deal with this.
-      return 'd';
+    // Specified format in axis/legend has higher precedence than fieldDef.format
+    if (specifiedFormat) {
+      return specifiedFormat;
     }
+
     // TODO: need to make this work correctly for numeric ordinal / nominal type
     return config.numberFormat;
   }
   return undefined;
 }
 
+function formatExpr(field: string, format: string) {
+  return `format(${field}, ${format ? `'${format}'` : null})`;
```

Comment:
idk, but https://github.com/vega/vega-lite/issues/1539#issuecomment-303504194 suggests that we should use null. It's a bit more explicit.